provider/oracle/images.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package oracle
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/juju/errors"
+	oci "github.com/juju/go-oracle-cloud/api"
+	"github.com/juju/utils/series"
+
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs/imagemetadata"
+	"github.com/juju/juju/environs/instances"
+)
+
+var windowsServerMap = map[string]string{
+	"Microsoft_Windows_Server_2012_R2": "win2012r2",
+	"Microsoft_Windows_Server_2008_R2": "win2018r2",
+}
+
+// instanceTypes returns all oracle cloud shapes and wraps them into instance.InstanceType
+// For more information about oracle cloud shapes, please see:
+// https://docs.oracle.com/cloud/latest/stcomputecs/STCSA/api-Shapes.html
+// https://docs.oracle.com/cloud/latest/stcomputecs/STCSG/GUID-1DD0FA71-AC7B-461C-B8C1-14892725AA69.htm#OCSUG210
+func instanceTypes(c *oci.Client) ([]instances.InstanceType, error) {
+	if c == nil {
+		return nil, errors.Errorf("cannot use nil client")
+	}
+
+	// fetch all shapes from the provider
+	shapes, err := c.AllShapes(nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// convert shapes to InstanceType
+	onlyArch := []string{"amd64"}
+	types := make([]instances.InstanceType, len(shapes.Result), len(shapes.Result))
+	for key, val := range shapes.Result {
+		types[key].Name = val.Name
+		types[key].Arches = onlyArch
+		types[key].Mem = val.Ram
+		types[key].CpuCores = uint64(val.Cpus)
+		types[key].RootDisk = val.Root_disk_size
+	}
+
+	return types, nil
+}
+
+// findInstanceSpec returns an *InstanceSpec, imagelist name
+// satisfying the supplied instanceConstraint
+func findInstanceSpec(
+	c *oci.Client,
+	allImageMetadata []*imagemetadata.ImageMetadata,
+	instanceType []instances.InstanceType,
+	ic *instances.InstanceConstraint,
+) (*instances.InstanceSpec, string, error) {
+
+	logger.Debugf("received %d image(s): %v", len(allImageMetadata), allImageMetadata)
+	version, err := series.SeriesVersion(ic.Series)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+	filtered := []*imagemetadata.ImageMetadata{}
+	for _, val := range allImageMetadata {
+		if val.Version != version {
+			continue
+		}
+		filtered = append(filtered, val)
+	}
+
+	images := instances.ImageMetadataToImages(filtered)
+	spec, err := instances.FindInstanceSpec(images, ic, instanceType)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	imagelist, err := getImageName(c, spec.Image.Id)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	return spec, imagelist, nil
+}
+
+func parseImageName(name string, uri *url.URL) (*imagemetadata.ImageMetadata, error) {
+	var id, arch, version string
+	if strings.HasPrefix(name, "Ubuntu") {
```

Comment:
These are offered by the oracle marketplace. The trend is as such. We will add a sample json returned by the API in one of the tests.